File: src/renderer/vulkan/renderpass/SwapchainPass.cpp
////////////////////////////////////////////////////////////////////////////////
/// Reaper
///
/// Copyright (c) 2015-2022 Thibault Schueller
/// This file is distributed under the MIT License
////////////////////////////////////////////////////////////////////////////////

#include "SwapchainPass.h"

#include "FrameGraphPass.h"
#include "ShadowConstants.h"

#include "renderer/graph/FrameGraphBuilder.h"
#include "renderer/vulkan/Backend.h"
#include "renderer/vulkan/CommandBuffer.h"
#include "renderer/vulkan/DescriptorSet.h"
#include "renderer/vulkan/FrameGraphResources.h"
#include "renderer/vulkan/GpuProfile.h"
#include "renderer/vulkan/Image.h"
#include "renderer/vulkan/Pipeline.h"
#include "renderer/vulkan/RenderPassHelpers.h"
#include "renderer/vulkan/SamplerResources.h"
#include "renderer/vulkan/ShaderModules.h"

#include "profiling/Scope.h"

#include "renderer/shader/swapchain_write.share.hlsl"

namespace Reaper
{
namespace
{
    struct SpecConstants
    {
        hlsl_uint color_space_index;
        hlsl_uint transfer_function_index;
        hlsl_uint dynamic_range;
    };

    u32 get_color_space_index(ColorSpace color_space)
    {
        switch (color_space)
        {
        case ColorSpace::sRGB:
            return COLOR_SPACE_SRGB;
        case ColorSpace::Rec709:
            return COLOR_SPACE_REC709;
        case ColorSpace::DisplayP3:
            return COLOR_SPACE_DISPLAY_P3;
        case ColorSpace::Rec2020:
            return COLOR_SPACE_REC2020;
        case ColorSpace::Unknown:
            AssertUnreachable();
        };

        return 0;
    }

    u32 get_transfer_function_index(TransferFunction func)
    {
        switch (func)
        {
        case TransferFunction::Linear:
            return TRANSFER_FUNC_LINEAR;
        case TransferFunction::sRGB:
            return TRANSFER_FUNC_SRGB;
        case TransferFunction::Rec709:
            return TRANSFER_FUNC_REC709;
        case TransferFunction::PQ:
            return TRANSFER_FUNC_PQ;
        case TransferFunction::scRGB_Windows:
            return TRANSFER_FUNC_WINDOWS_SCRGB;
        case TransferFunction::Unknown:
            AssertUnreachable();
        };

        return 0;
    }

    VkPipeline create_swapchain_pipeline(VulkanBackend& backend, const ShaderModules& shader_modules,
                                         VkPipelineLayout pipelineLayout)
    {
        std::vector<VkSpecializationMapEntry> specialization_constants_entries = {
            VkSpecializationMapEntry{
                0,
                offsetof(SpecConstants, color_space_index),
                sizeof(SpecConstants::color_space_index),
            },
            VkSpecializationMapEntry{
                1,
                offsetof(SpecConstants, transfer_function_index),
                sizeof(SpecConstants::transfer_function_index),
            },
            VkSpecializationMapEntry{
                2,
                offsetof(SpecConstants, dynamic_range),
                sizeof(SpecConstants::dynamic_range),
            },
        };

        const SwapchainFormat& swapchain_format = backend.presentInfo.swapchain_format;

        SpecConstants spec_constants;
        spec_constants.color_space_index = get_color_space_index(swapchain_format.color_space);
        spec_constants.transfer_function_index = get_transfer_function_index(swapchain_format.transfer_function);
        spec_constants.dynamic_range = swapchain_format.is_hdr ? DYNAMIC_RANGE_HDR : DYNAMIC_RANGE_SDR;

        VkSpecializationInfo specialization = {
            static_cast<u32>(specialization_constants_entries.size()), // uint32_t mapEntryCount;
            specialization_constants_entries.data(), // const VkSpecializationMapEntry*    pMapEntries;
            sizeof(SpecConstants),                   // size_t                             dataSize;
            &spec_constants,                         // const void*                        pData;
        };

        const VkShaderModuleCreateInfo module_create_info_vert =
            shader_module_create_info(get_spirv_shader_module(shader_modules, "fullscreen_triangle.vert.spv"));
        const VkShaderModuleCreateInfo module_create_info_frag =
            shader_module_create_info(get_spirv_shader_module(shader_modules, "swapchain_write.frag.spv"));

        std::vector<VkPipelineShaderStageCreateInfo> shader_stages = {
            default_pipeline_shader_stage_create_info(VK_SHADER_STAGE_VERTEX_BIT, &module_create_info_vert),
            default_pipeline_shader_stage_create_info(VK_SHADER_STAGE_FRAGMENT_BIT, &module_create_info_frag,
                                                      &specialization)};

        VkPipelineColorBlendAttachmentState blend_attachment_state = default_pipeline_color_blend_attachment_state();

        GraphicsPipelineProperties pipeline_properties = default_graphics_pipeline_properties();
        pipeline_properties.blend_state.attachmentCount = 1;
        pipeline_properties.blend_state.pAttachments = &blend_attachment_state;
        pipeline_properties.pipeline_layout = pipelineLayout;
        pipeline_properties.pipeline_rendering.colorAttachmentCount = 1;
        pipeline_properties.pipeline_rendering.pColorAttachmentFormats = &swapchain_format.vk_view_format;

        std::vector<VkDynamicState> dynamic_states = {VK_DYNAMIC_STATE_VIEWPORT, VK_DYNAMIC_STATE_SCISSOR};

        VkPipeline pipeline =
            create_graphics_pipeline(backend.device, shader_stages, pipeline_properties, dynamic_states);

        Assert(backend.physical_device.graphics_queue_family_index
               == backend.physical_device.present_queue_family_index);

        return pipeline;
    }
} // namespace

SwapchainPassResources create_swapchain_pass_resources(VulkanBackend& backend, const ShaderModules& shader_modules)
{
    SwapchainPassResources resources = {};

    std::vector<VkDescriptorSetLayoutBinding> descriptorSetLayoutBinding = {
        {0, VK_DESCRIPTOR_TYPE_SAMPLER, 1, VK_SHADER_STAGE_FRAGMENT_BIT, nullptr},
        {1, VK_DESCRIPTOR_TYPE_SAMPLED_IMAGE, 1, VK_SHADER_STAGE_FRAGMENT_BIT, nullptr},
        {2, VK_DESCRIPTOR_TYPE_SAMPLED_IMAGE, 1, VK_SHADER_STAGE_FRAGMENT_BIT, nullptr},
        {3, VK_DESCRIPTOR_TYPE_SAMPLED_IMAGE, 1, VK_SHADER_STAGE_FRAGMENT_BIT, nullptr},
        {4, VK_DESCRIPTOR_TYPE_STORAGE_BUFFER, 1, VK_SHADER_STAGE_FRAGMENT_BIT, nullptr},
        {5, VK_DESCRIPTOR_TYPE_SAMPLED_IMAGE, 1, VK_SHADER_STAGE_FRAGMENT_BIT, nullptr},
        {6, VK_DESCRIPTOR_TYPE_SAMPLED_IMAGE, 1, VK_SHADER_STAGE_FRAGMENT_BIT, nullptr},
    };

    resources.descriptorSetLayout = create_descriptor_set_layout(backend.device, descriptorSetLayoutBinding);

    const VkPushConstantRange push_constant_range = {VK_SHADER_STAGE_FRAGMENT_BIT, 0, sizeof(SwapchainWriteParams)};

    resources.pipelineLayout = create_pipeline_layout(backend.device, std::span(&resources.descriptorSetLayout, 1),
                                                      std::span(&push_constant_range, 1));
    resources.pipeline = create_swapchain_pipeline(backend, shader_modules, resources.pipelineLayout);

    allocate_descriptor_sets(backend.device, backend.global_descriptor_pool,
                             std::span(&resources.descriptorSetLayout, 1), std::span(&resources.descriptor_set, 1));

    return resources;
}

void destroy_swapchain_pass_resources(VulkanBackend& backend, const SwapchainPassResources& resources)
{
    vkDestroyPipeline(backend.device, resources.pipeline, nullptr);
    vkDestroyPipelineLayout(backend.device, resources.pipelineLayout, nullptr);
    vkDestroyDescriptorSetLayout(backend.device, resources.descriptorSetLayout, nullptr);
}

void reload_swapchain_pipeline(VulkanBackend& backend, const ShaderModules& shader_modules,
                               SwapchainPassResources& resources)
{
    REAPER_PROFILE_SCOPE_FUNC();

    vkDestroyPipeline(backend.device, resources.pipeline, nullptr);

    resources.pipeline = create_swapchain_pipeline(backend, shader_modules, resources.pipelineLayout);
}

SwapchainFrameGraphRecord
create_swapchain_pass_record(FrameGraph::Builder&            builder,
                             FrameGraph::ResourceUsageHandle scene_hdr_usage_handle,
                             FrameGraph::ResourceUsageHandle split_tiled_lighting_hdr_usage_handle,
                             FrameGraph::ResourceUsageHandle gui_sdr_usage_handle,
                             FrameGraph::ResourceUsageHandle histogram_buffer_usage_handle,
                             FrameGraph::ResourceUsageHandle average_exposure_usage_handle,
                             FrameGraph::ResourceUsageHandle tone_map_lut,
                             FrameGraph::ResourceUsageHandle tiled_debug_texture_overlay_usage_handle)
{
    SwapchainFrameGraphRecord swapchain;
    swapchain.pass_handle = builder.create_render_pass("Swapchain", true);

    swapchain.scene_hdr = builder.read_texture(swapchain.pass_handle, scene_hdr_usage_handle,
                                               GPUTextureAccess{.stage_mask = VK_PIPELINE_STAGE_2_FRAGMENT_SHADER_BIT,
                                                                .access_mask = VK_ACCESS_2_SHADER_READ_BIT,
                                                                .image_layout = VK_IMAGE_LAYOUT_READ_ONLY_OPTIMAL});

    swapchain.lighting_result =
        builder.read_texture(swapchain.pass_handle, split_tiled_lighting_hdr_usage_handle,
                             GPUTextureAccess{VK_PIPELINE_STAGE_2_FRAGMENT_SHADER_BIT, VK_ACCESS_2_SHADER_READ_BIT,
                                              VK_IMAGE_LAYOUT_READ_ONLY_OPTIMAL});

    swapchain.gui =
        builder.read_texture(swapchain.pass_handle, gui_sdr_usage_handle,
                             GPUTextureAccess{VK_PIPELINE_STAGE_2_FRAGMENT_SHADER_BIT, VK_ACCESS_2_SHADER_READ_BIT,
                                              VK_IMAGE_LAYOUT_READ_ONLY_OPTIMAL});

    // FIXME just to hook the pass to the render graph
    swapchain.histogram =
        builder.read_buffer(swapchain.pass_handle, histogram_buffer_usage_handle,
                            GPUBufferAccess{VK_PIPELINE_STAGE_2_FRAGMENT_SHADER_BIT, VK_ACCESS_2_SHADER_READ_BIT});

    swapchain.average_exposure =
        builder.read_buffer(swapchain.pass_handle, average_exposure_usage_handle,
                            GPUBufferAccess{VK_PIPELINE_STAGE_2_FRAGMENT_SHADER_BIT, VK_ACCESS_2_SHADER_READ_BIT});

    swapchain.tone_map_lut =
        builder.read_texture(swapchain.pass_handle, tone_map_lut,
                             GPUTextureAccess{VK_PIPELINE_STAGE_2_FRAGMENT_SHADER_BIT, VK_ACCESS_2_SHADER_READ_BIT,
                                              VK_IMAGE_LAYOUT_READ_ONLY_OPTIMAL});

    swapchain.tile_debug =
        builder.read_texture(swapchain.pass_handle, tiled_debug_texture_overlay_usage_handle,
                             GPUTextureAccess{VK_PIPELINE_STAGE_2_FRAGMENT_SHADER_BIT, VK_ACCESS_2_SHADER_READ_BIT,
                                              VK_IMAGE_LAYOUT_READ_ONLY_OPTIMAL});

    return swapchain;
}

void update_swapchain_pass_descriptor_set(const FrameGraph::FrameGraph&    frame_graph,
                                          const FrameGraphResources&       frame_graph_resources,
                                          const SwapchainFrameGraphRecord& record, DescriptorWriteHelper& write_helper,
                                          const SwapchainPassResources& resources,
                                          const SamplerResources&       sampler_resources)
{
    const FrameGraphTexture hdr_scene_texture =
        get_frame_graph_texture(frame_graph_resources, frame_graph, record.scene_hdr);
    const FrameGraphTexture lighting_texture =
        get_frame_graph_texture(frame_graph_resources, frame_graph, record.lighting_result);
    const FrameGraphTexture gui_texture = get_frame_graph_texture(frame_graph_resources, frame_graph, record.gui);
    const FrameGraphBuffer  average_exposure =
        get_frame_graph_buffer(frame_graph_resources, frame_graph, record.average_exposure);
    const FrameGraphTexture tile_lighting_debug_texture =
        get_frame_graph_texture(frame_graph_resources, frame_graph, record.tile_debug);
    const FrameGraphTexture tone_map_lut =
        get_frame_graph_texture(frame_graph_resources, frame_graph, record.tone_map_lut);

    write_helper.append(resources.descriptor_set, 0, sampler_resources.linear_black_border);
    write_helper.append(resources.descriptor_set, 1, VK_DESCRIPTOR_TYPE_SAMPLED_IMAGE,
                        hdr_scene_texture.default_view_handle, hdr_scene_texture.image_layout);
    write_helper.append(resources.descriptor_set, 2, VK_DESCRIPTOR_TYPE_SAMPLED_IMAGE,
                        lighting_texture.default_view_handle, lighting_texture.image_layout);
    write_helper.append(resources.descriptor_set, 3, VK_DESCRIPTOR_TYPE_SAMPLED_IMAGE, gui_texture.default_view_handle,
                        gui_texture.image_layout);
    write_helper.append(resources.descriptor_set, 4, VK_DESCRIPTOR_TYPE_STORAGE_BUFFER, average_exposure.handle);
    write_helper.append(resources.descriptor_set, 5, VK_DESCRIPTOR_TYPE_SAMPLED_IMAGE, tone_map_lut.default_view_handle,
                        tone_map_lut.image_layout);
    write_helper.append(resources.descriptor_set, 6, VK_DESCRIPTOR_TYPE_SAMPLED_IMAGE,
                        tile_lighting_debug_texture.default_view_handle, tile_lighting_debug_texture.image_layout);
}

void record_swapchain_command_buffer(const FrameGraphHelper&          frame_graph_helper,
                                     const SwapchainFrameGraphRecord& pass_record, CommandBuffer& cmdBuffer,
                                     const SwapchainPassResources& pass_resources, VkImageView swapchain_buffer_view,
                                     VkExtent2D swapchain_extent, float exposure_compensation_stops,
                                     float tonemap_min_nits, float tonemap_max_nits, float sdr_ui_max_brightness_nits,
                                     float sdr_peak_brightness_nits)
{
    REAPER_GPU_SCOPE(cmdBuffer, "Swapchain");

    const FrameGraphBarrierScope framegraph_barrier_scope(cmdBuffer, frame_graph_helper, pass_record.pass_handle);

    const VkRect2D   pass_rect = default_vk_rect(swapchain_extent);
    const VkViewport viewport = default_vk_viewport(pass_rect);

    vkCmdBindPipeline(cmdBuffer.handle, VK_PIPELINE_BIND_POINT_GRAPHICS, pass_resources.pipeline);

    vkCmdSetViewport(cmdBuffer.handle, 0, 1, &viewport);
    vkCmdSetScissor(cmdBuffer.handle, 0, 1, &pass_rect);

    VkRenderingAttachmentInfo color_attachment =
        default_rendering_attachment_info(swapchain_buffer_view, VK_IMAGE_LAYOUT_ATTACHMENT_OPTIMAL);
    color_attachment.loadOp = VK_ATTACHMENT_LOAD_OP_DONT_CARE;

    const VkRenderingInfo rendering_info = default_rendering_info(pass_rect, &color_attachment);

    vkCmdBeginRendering(cmdBuffer.handle, &rendering_info);

    SwapchainWriteParams push_constants;
    push_constants.exposure_compensation = exp2(exposure_compensation_stops);
    push_constants.tonemap_min_nits = tonemap_min_nits;
    push_constants.tonemap_max_nits = tonemap_max_nits;
    push_constants.sdr_ui_max_brightness_nits = sdr_ui_max_brightness_nits;
    push_constants.sdr_peak_brightness_nits = sdr_peak_brightness_nits;

    vkCmdPushConstants(cmdBuffer.handle, pass_resources.pipelineLayout, VK_SHADER_STAGE_FRAGMENT_BIT, 0,
                       sizeof(push_constants), &push_constants);

    vkCmdBindDescriptorSets(cmdBuffer.handle, VK_PIPELINE_BIND_POINT_GRAPHICS, pass_resources.pipelineLayout, 0, 1,
                            &pass_resources.descriptor_set, 0, nullptr);

    vkCmdDraw(cmdBuffer.handle, 3, 1, 0, 0);

    vkCmdEndRendering(cmdBuffer.handle);
}
} // namespace Reaper

File: src/renderer/vulkan/renderpass/swapchain_pass.zig
// Port of src/renderer/vulkan/renderpass/SwapchainPass.{h,cpp}
//
// The final composite: tone maps the HDR scene through the baked LUT and writes
// the swapchain image. The colour space, transfer function and dynamic range
// are specialization constants baked from the swapchain format, so the pipeline
// has to be rebuilt whenever the swapchain is reconfigured — which is why it
// lives outside the pipeline factory.
//
// The shader wants seven inputs. Four of them come from passes that do not
// exist yet (tiled lighting in M6, GUI/histogram/exposure in M7), so those are
// fed placeholder resources declared in the graph. They are real graph
// resources rather than unbound descriptors: the fragment shader samples all of
// them unconditionally, so leaving them unbound would be undefined behaviour
// rather than simply "off".

const std = @import("std");
const vk = @import("vulkan");

const buffer_module = @import("../buffer.zig");
const descriptor_set = @import("../descriptor_set.zig");
const fg = @import("../../graph/frame_graph.zig");
const frame_graph_pass = @import("frame_graph_pass.zig");
const gpu_buffer = @import("../../buffer/gpu_buffer.zig");
const gpu_texture_properties = @import("../../texture/gpu_texture_properties.zig");
const pipeline_module = @import("../pipeline.zig");
const render_pass_helpers = @import("../render_pass_helpers.zig");
const shader_modules = @import("../shader_modules.zig");

const hlsl = @import("../../hlsl/swapchain_write.zig");

const Builder = @import("../../graph/builder.zig").Builder;
const DescriptorWriteHelper = descriptor_set.DescriptorWriteHelper;
const FrameGraphResources = @import("../framegraph_resources.zig").FrameGraphResources;
const SamplerResources = @import("../sampler_resources.zig").SamplerResources;
const Swapchain = @import("../Swapchain.zig");

// --------------------------------------------------------------------------
// Specialization constants
// --------------------------------------------------------------------------

const SpecConstants = extern struct {
    color_space_index: u32,
    transfer_function_index: u32,
    dynamic_range: u32,
};

fn getColorSpaceIndex(color_space: Swapchain.ColorSpace) u32 {
    return switch (color_space) {
        .sRGB => hlsl.COLOR_SPACE_SRGB,
        .Rec709 => hlsl.COLOR_SPACE_REC709,
        .DisplayP3 => hlsl.COLOR_SPACE_DISPLAY_P3,
        .Rec2020 => hlsl.COLOR_SPACE_REC2020,
        .Unknown => unreachable,
    };
}

fn getTransferFunctionIndex(transfer_function: Swapchain.TransferFunction) u32 {
    return switch (transfer_function) {
        .Linear => hlsl.TRANSFER_FUNC_LINEAR,
        .sRGB => hlsl.TRANSFER_FUNC_SRGB,
        .Rec709 => hlsl.TRANSFER_FUNC_REC709,
        .PQ => hlsl.TRANSFER_FUNC_PQ,
        .scRGB_Windows => hlsl.TRANSFER_FUNC_WINDOWS_SCRGB,
        .Unknown => unreachable,
    };
}

// --------------------------------------------------------------------------
// Resources
// --------------------------------------------------------------------------

pub const SwapchainPassResources = struct {
    descriptor_set_layout: vk.DescriptorSetLayout,
    pipeline_layout: vk.PipelineLayout,
    pipeline: vk.Pipeline,
    descriptor_set: vk.DescriptorSet,

    /// The swapchain format the pipeline's spec constants were baked from.
    swapchain_format: Swapchain.SwapchainFormat,

    pub fn init(
        vkd: anytype,
        device: vk.Device,
        descriptor_pool: vk.DescriptorPool,
        swapchain_format: Swapchain.SwapchainFormat,
    ) !SwapchainPassResources {
        const bindings = [_]vk.DescriptorSetLayoutBinding{
            .{ .binding = 0, .descriptor_type = .sampler, .descriptor_count = 1, .stage_flags = .{ .fragment_bit = true }, .p_immutable_samplers = null },
            .{ .binding = 1, .descriptor_type = .sampled_image, .descriptor_count = 1, .stage_flags = .{ .fragment_bit = true }, .p_immutable_samplers = null },
            .{ .binding = 2, .descriptor_type = .sampled_image, .descriptor_count = 1, .stage_flags = .{ .fragment_bit = true }, .p_immutable_samplers = null },
            .{ .binding = 3, .descriptor_type = .sampled_image, .descriptor_count = 1, .stage_flags = .{ .fragment_bit = true }, .p_immutable_samplers = null },
            .{ .binding = 4, .descriptor_type = .storage_buffer, .descriptor_count = 1, .stage_flags = .{ .fragment_bit = true }, .p_immutable_samplers = null },
            .{ .binding = 5, .descriptor_type = .sampled_image, .descriptor_count = 1, .stage_flags = .{ .fragment_bit = true }, .p_immutable_samplers = null },
            .{ .binding = 6, .descriptor_type = .sampled_image, .descriptor_count = 1, .stage_flags = .{ .fragment_bit = true }, .p_immutable_samplers = null },
        };

        const descriptor_set_layout = try pipeline_module.createDescriptorSetLayout(vkd, device, &bindings, &.{});
        errdefer vkd.destroyDescriptorSetLayout(device, descriptor_set_layout, null);

        const push_constant_ranges = [_]vk.PushConstantRange{.{
            .stage_flags = .{ .fragment_bit = true },
            .offset = 0,
            .size = @sizeOf(hlsl.SwapchainWriteParams),
        }};

        const pipeline_layout = try pipeline_module.createPipelineLayout(
            vkd,
            device,
            &.{descriptor_set_layout},
            &push_constant_ranges,
        );
        errdefer vkd.destroyPipelineLayout(device, pipeline_layout, null);

        const pipeline = try createPipeline(vkd, device, pipeline_layout, swapchain_format);
        errdefer vkd.destroyPipeline(device, pipeline, null);

        var sets: [1]vk.DescriptorSet = undefined;
        try pipeline_module.allocateDescriptorSets(vkd, device, descriptor_pool, &.{descriptor_set_layout}, &sets);

        return .{
            .descriptor_set_layout = descriptor_set_layout,
            .pipeline_layout = pipeline_layout,
            .pipeline = pipeline,
            .descriptor_set = sets[0],
            .swapchain_format = swapchain_format,
        };
    }

    pub fn deinit(self: *SwapchainPassResources, vkd: anytype, device: vk.Device) void {
        vkd.destroyPipeline(device, self.pipeline, null);
        vkd.destroyPipelineLayout(device, self.pipeline_layout, null);
        vkd.destroyDescriptorSetLayout(device, self.descriptor_set_layout, null);
    }

    /// Mirrors reload_swapchain_pipeline(). The caller must have waited for the
    /// device to go idle.
    pub fn reconfigure(
        self: *SwapchainPassResources,
        vkd: anytype,
        device: vk.Device,
        swapchain_format: Swapchain.SwapchainFormat,
    ) !void {
        if (std.meta.eql(self.swapchain_format, swapchain_format)) return;

        const pipeline = try createPipeline(vkd, device, self.pipeline_layout, swapchain_format);

        vkd.destroyPipeline(device, self.pipeline, null);
        self.pipeline = pipeline;
        self.swapchain_format = swapchain_format;
    }
};

fn createPipeline(
    vkd: anytype,
    device: vk.Device,
    pipeline_layout: vk.PipelineLayout,
    swapchain_format: Swapchain.SwapchainFormat,
) !vk.Pipeline {
    const spec_constants = SpecConstants{
        .color_space_index = getColorSpaceIndex(swapchain_format.color_space),
        .transfer_function_index = getTransferFunctionIndex(swapchain_format.transfer_function),
        .dynamic_range = if (swapchain_format.is_hdr) hlsl.DYNAMIC_RANGE_HDR else hlsl.DYNAMIC_RANGE_SDR,
    };

    const spec_entries = [_]vk.SpecializationMapEntry{
        .{ .constant_id = 0, .offset = @offsetOf(SpecConstants, "color_space_index"), .size = @sizeOf(u32) },
        .{ .constant_id = 1, .offset = @offsetOf(SpecConstants, "transfer_function_index"), .size = @sizeOf(u32) },
        .{ .constant_id = 2, .offset = @offsetOf(SpecConstants, "dynamic_range"), .size = @sizeOf(u32) },
    };

    const specialization = vk.SpecializationInfo{
        .map_entry_count = spec_entries.len,
        .p_map_entries = &spec_entries,
        .data_size = @sizeOf(SpecConstants),
        .p_data = &spec_constants,
    };

    const module_create_info_vert = pipeline_module.shaderModuleCreateInfo(
        shader_modules.get("fullscreen_triangle.vert.spv"),
    );
    const module_create_info_frag = pipeline_module.shaderModuleCreateInfo(
        shader_modules.get("swapchain_write.frag.spv"),
    );

    const shader_stages = [_]vk.PipelineShaderStageCreateInfo{
        pipeline_module.defaultPipelineShaderStageCreateInfo(.{ .vertex_bit = true }, &module_create_info_vert, null),
        pipeline_module.defaultPipelineShaderStageCreateInfo(
            .{ .fragment_bit = true },
            &module_create_info_frag,
            &specialization,
        ),
    };

    const blend_attachment_states = [_]vk.PipelineColorBlendAttachmentState{
        pipeline_module.defaultPipelineColorBlendAttachmentState(),
    };

    const color_formats = [_]vk.Format{swapchain_format.vk_view_format};

    var properties = pipeline_module.defaultGraphicsPipelineProperties(null);
    properties.blend_state.attachment_count = blend_attachment_states.len;
    properties.blend_state.p_attachments = &blend_attachment_states;
    properties.pipeline_layout = pipeline_layout;
    properties.pipeline_rendering.color_attachment_count = color_formats.len;
    properties.pipeline_rendering.p_color_attachment_formats = &color_formats;

    const dynamic_states = [_]vk.DynamicState{ .viewport, .scissor };

    return pipeline_module.createGraphicsPipeline(vkd, device, &shader_stages, &properties, &dynamic_states);
}

// --------------------------------------------------------------------------
// Frame graph record
// --------------------------------------------------------------------------

pub const SwapchainFrameGraphRecord = struct {
    pass_handle: fg.RenderPassHandle,
    scene_hdr: fg.ResourceUsageHandle,
    lighting_result: fg.ResourceUsageHandle,
    gui: fg.ResourceUsageHandle,
    histogram: fg.ResourceUsageHandle,
    average_exposure: fg.ResourceUsageHandle,
    tone_map_lut: fg.ResourceUsageHandle,
    tile_debug: fg.ResourceUsageHandle,
};

const read_sampled = fg.GPUResourceAccess{
    .stage_mask = .{ .fragment_shader_bit = true },
    .access_mask = .{ .shader_read_bit = true },
    .image_layout = .read_only_optimal,
};

const read_storage = fg.GPUResourceAccess{
    .stage_mask = .{ .fragment_shader_bit = true },
    .access_mask = .{ .shader_read_bit = true },
    .image_layout = .undefined,
};

/// What the composite samples besides the forward pass's HDR target. Every one
/// of these now comes from a real pass.
pub const SwapchainInputs = struct {
    lighting_result: fg.ResourceUsageHandle,
    gui: fg.ResourceUsageHandle,
    histogram: fg.ResourceUsageHandle,
    average_exposure: fg.ResourceUsageHandle,
    tile_debug: fg.ResourceUsageHandle,
};

pub fn createFrameGraphRecord(
    builder: *Builder,
    scene_hdr_usage_handle: fg.ResourceUsageHandle,
    inputs: SwapchainInputs,
    tone_map_lut: fg.ResourceUsageHandle,
) !SwapchainFrameGraphRecord {
    const pass_handle = try builder.createRenderPass("Swapchain", true);

    return .{
        .pass_handle = pass_handle,
        .scene_hdr = try builder.readTexture(pass_handle, scene_hdr_usage_handle, fg.toTextureAccess(read_sampled), &.{}),
        .lighting_result = try builder.readTexture(pass_handle, inputs.lighting_result, fg.toTextureAccess(read_sampled), &.{}),
        .gui = try builder.readTexture(pass_handle, inputs.gui, fg.toTextureAccess(read_sampled), &.{}),
        // FIXME just to hook the pass to the render graph
        .histogram = try builder.readBuffer(pass_handle, inputs.histogram, fg.toBufferAccess(read_storage), &.{}),
        .average_exposure = try builder.readBuffer(pass_handle, inputs.average_exposure, fg.toBufferAccess(read_storage), &.{}),
        .tone_map_lut = try builder.readTexture(pass_handle, tone_map_lut, fg.toTextureAccess(read_sampled), &.{}),
        .tile_debug = try builder.readTexture(pass_handle, inputs.tile_debug, fg.toTextureAccess(read_sampled), &.{}),
    };
}

// --------------------------------------------------------------------------
// Descriptor updates and recording
// --------------------------------------------------------------------------

pub fn updateDescriptorSet(
    write_helper: *DescriptorWriteHelper,
    framegraph: *const fg.FrameGraph,
    frame_graph_resources: *const FrameGraphResources,
    record: SwapchainFrameGraphRecord,
    resources: *const SwapchainPassResources,
    sampler_resources: SamplerResources,
) void {
    const hdr_scene_texture = frame_graph_resources.getTexture(framegraph, record.scene_hdr);
    const lighting_texture = frame_graph_resources.getTexture(framegraph, record.lighting_result);
    const gui_texture = frame_graph_resources.getTexture(framegraph, record.gui);
    const average_exposure = frame_graph_resources.getBuffer(framegraph, record.average_exposure);
    const tile_debug_texture = frame_graph_resources.getTexture(framegraph, record.tile_debug);
    const tone_map_lut = frame_graph_resources.getTexture(framegraph, record.tone_map_lut);

    const set = resources.descriptor_set;

    write_helper.appendSampler(set, 0, sampler_resources.linear_black_border);
    write_helper.appendImage(set, 1, .sampled_image, hdr_scene_texture.default_view_handle, hdr_scene_texture.image_layout);
    write_helper.appendImage(set, 2, .sampled_image, lighting_texture.default_view_handle, lighting_texture.image_layout);
    write_helper.appendImage(set, 3, .sampled_image, gui_texture.default_view_handle, gui_texture.image_layout);
    write_helper.appendBuffer(set, 4, .storage_buffer, average_exposure.handle, 0, vk.WHOLE_SIZE);
    write_helper.appendImage(set, 5, .sampled_image, tone_map_lut.default_view_handle, tone_map_lut.image_layout);
    write_helper.appendImage(set, 6, .sampled_image, tile_debug_texture.default_view_handle, tile_debug_texture.image_layout);
}

pub const CompositeParams = struct {
    exposure_compensation_stops: f32,
    tonemap_min_nits: f32,
    tonemap_max_nits: f32,
    sdr_ui_max_brightness_nits: f32,
    sdr_peak_brightness_nits: f32,
};

pub fn recordCommandBuffer(
    vkd: anytype,
    cmd_buffer: vk.CommandBuffer,
    helper: frame_graph_pass.FrameGraphHelper,
    record: SwapchainFrameGraphRecord,
    resources: *const SwapchainPassResources,
    swapchain_buffer_view: vk.ImageView,
    swapchain_extent: vk.Extent2D,
    params: CompositeParams,
) void {
    frame_graph_pass.beginBarrierScope(vkd, cmd_buffer, helper, record.pass_handle);
    defer frame_graph_pass.endBarrierScope(vkd, cmd_buffer, helper, record.pass_handle);

    const pass_rect = render_pass_helpers.defaultRect(swapchain_extent);
    const viewports = [_]vk.Viewport{render_pass_helpers.defaultViewport(pass_rect)};
    const scissors = [_]vk.Rect2D{pass_rect};

    vkd.cmdBindPipeline(cmd_buffer, .graphics, resources.pipeline);

    vkd.cmdSetViewport(cmd_buffer, 0, &viewports);
    vkd.cmdSetScissor(cmd_buffer, 0, &scissors);

    var color_attachments = [_]vk.RenderingAttachmentInfo{
        pipeline_module.defaultRenderingAttachmentInfo(swapchain_buffer_view, .attachment_optimal),
    };
    color_attachments[0].load_op = .dont_care;

    const rendering_info = pipeline_module.defaultRenderingInfo(pass_rect, &color_attachments, null);

    vkd.cmdBeginRendering(cmd_buffer, &rendering_info);

    const push_constants = hlsl.SwapchainWriteParams{
        .exposure_compensation = std.math.exp2(params.exposure_compensation_stops),
        .tonemap_min_nits = params.tonemap_min_nits,
        .tonemap_max_nits = params.tonemap_max_nits,
        .sdr_ui_max_brightness_nits = params.sdr_ui_max_brightness_nits,
        .sdr_peak_brightness_nits = params.sdr_peak_brightness_nits,
    };

    vkd.cmdPushConstants(
        cmd_buffer,
        resources.pipeline_layout,
        .{ .fragment_bit = true },
        0,
        @sizeOf(hlsl.SwapchainWriteParams),
        &push_constants,
    );

    const sets = [_]vk.DescriptorSet{resources.descriptor_set};
    vkd.cmdBindDescriptorSets(cmd_buffer, .graphics, resources.pipeline_layout, 0, &sets, &.{});

    vkd.cmdDraw(cmd_buffer, 3, 1, 0, 0);

    vkd.cmdEndRendering(cmd_buffer);
}

// --------------------------------------------------------------------------
// Tests
// --------------------------------------------------------------------------

const testing = std.testing;

test "spec constant offsets match the shader's constant ids" {
    // The shader declares VK_CONSTANT(0..2) in this order; a reordering here
    // would silently tone map with the wrong transfer function.
    try testing.expectEqual(@as(usize, 0), @offsetOf(SpecConstants, "color_space_index"));
    try testing.expectEqual(@as(usize, 4), @offsetOf(SpecConstants, "transfer_function_index"));
    try testing.expectEqual(@as(usize, 8), @offsetOf(SpecConstants, "dynamic_range"));
    try testing.expectEqual(@as(usize, 12), @sizeOf(SpecConstants));
}

test "swapchain format maps to the shader's enums" {
    try testing.expectEqual(hlsl.COLOR_SPACE_SRGB, getColorSpaceIndex(.sRGB));
    try testing.expectEqual(hlsl.COLOR_SPACE_REC2020, getColorSpaceIndex(.Rec2020));

    // An sRGB swapchain has its EOTF applied by the image view, so the shader
    // is told the transfer function is linear — see configureVulkanWmSwapchain.
    try testing.expectEqual(hlsl.TRANSFER_FUNC_LINEAR, getTransferFunctionIndex(.Linear));
    try testing.expectEqual(hlsl.TRANSFER_FUNC_PQ, getTransferFunctionIndex(.PQ));
}
